### ISSUE TITLE
Wrong determination of redundant rows

### ISSUE THREAD

**Issue by kliem** [cddlib/cddlib#45]
For some reason rows 1,3 and 5 are considered redundant, but row 1 isn't.

This is version 0.94j as included in SageMath and reported as [https://trac.sagemath.org/ticket/30319]( https://trac.sagemath.org/ticket/30319).

    ---- CDD input -----
    H-representation
    linearity 2 1 2
    begin
     5 4 real
     0.0 0.0 1.0 1.0
     -1.0 0.0 1.0 1.0
     -1.0 2.0 -1.0 0.0
     -1.0 4.0 -2.0 0.0
     1 0 0 0
    end
    
    ---- CDD output -----
    Canonicalize the matrix.
    Implicit linearity rows are:
    
    Redundant rows are:1 3 5 
    
    Nonredundant representation:
    The new row positions are as follows (orig:new).
    Each redundant row has the new number 0.
    Each deleted duplicated row has a number nagative of the row that
    represents its equivalence class.
     1:0 2:1 3:0 4:2 5:0
    H-representation
    linearity 1  1
    begin
     2 4 real
     -1  0  1  1
     -1  4 -2  0
    end
    
    size = 5 x 4
    Number Type = real

**Comment by dimpase** [comment]
As discussed on https://trac.sagemath.org/ticket/30319 it also happens with cdd on rational data, not only on RDF.

**Comment by saraedum** [comment]
@komeifukuda could you have a look at this one?

I am not actually all that familiar with the code base so if someone volunteers to have a look that would be great.

**Comment by komeif** [comment]
@dimpase You are right.  This polytope is of course the empty set.  A nonredundant representation of the empty set is not unique (in fact infinite), and somehow cddlib chose a wrong set of inequalities to represent the empty set.  I will check it out.  Thank you for the bug report.  Komei

**Comment by komeif** [comment]
@dimpase @saraedum Here is the fix.  On the line 2640 (the function dd_MatrixRedundancyRemove) in cddlp.c,
please replace
if (dd_RedundantExtensive(M1, k, cvec, &redset1,error)) {
with
if (dd_Redundant(M1, k, cvec, error)) {

I still need to check whether the function dd_RedundantExtensive is working correctly.
This is an independent issue that I will keep working on.

Another thing, we must reflect the version number of cddlib in the file cddtypes.h .
Namely the line
#define dd_DDVERSION   "Version 0.94k (September 26, 2020)"
must be updated.  Above, I put today's date for the new version, but this should be
modified each time according to the date of a new release.

Thank you for correcting this for the version 0.94k.  Komei

**Comment by komeif** [comment]
@dimpase @saraedum  Now, I see a better way to fix the bug.  The best is to fix 
dd_RedundantExtensive.  In this function cddlp.c (at lines 2538-2541),
---
   set_copy(*redset,lp->redset_extra);
    set_delelem(*redset, itest);
    /* itest row might be redundant in the lp but this has nothing to do with its redundancy
    in the original system M.   Thus we must delete it.  */
---
we need to add one line (set_diff line) as
---
    set_copy(*redset,lp->redset_extra);
    set_diff(*redset, *redset, M->linset);  /* linearity set is not tested for redundancy */
    set_delelem(*redset, itest);
    /* itest row might be redundant in the lp but this has nothing to do with its redundancy
    in the original system M.   Thus we must delete it.  */
---
The  bug was not excluding the linearity set from the redundant set.

My previous suggestion must be ignored, because this fix of dd_RedundantExtensive
results in a more efficient recognition of redundancy.

I hope that my correction is clear.

Komei

**Comment by slel** [comment]
@komeif can you send a pull request with the fix?

**Comment by komeif** [comment]
@dimpase I wonder if you can send a pull request for me.  The change is so little (adding one line in dd_RedundantExtensive).  I really do not understand how GitHub works, and most likely I will do stupid things if I try to do this request.  Of course I take the full responsibility of this change.

**Comment by kliem** [comment]
I created https://github.com/cddlib/cddlib/pull/54

**Comment by dimpase** [comment]
@komeif - is this what you kept in mind, cf. #54 ?

**Comment by komeif** [comment]
Yes, thank you very much.

**Comment by dimpase** [comment]
I'm trying to test #54, but already the current master 879960a7c6c1a913aeffa0002008b000838472ec does not seem to reproduce the error. Specifically, I'm doing on, x86_64 Linux
```
$ ./bootstrap
$ ./configure && make && make install
$ cddexec --rep <examples/bug.ine 
size = 5 x 4
Number Type = real
The second representation:
V-representation
begin
 0 4 real
end
```
with
```
$ cat examples/bug.ine
H-representation
linearity 2 1 2
begin
 5 4 real
 0.0 0.0 1.0 1.0
 -1.0 0.0 1.0 1.0
 -1.0 2.0 -1.0 0.0
 -1.0 4.0 -2.0 0.0
 1 0 0 0
end
```

Am I doing something wrong?

**Comment by komeif** [comment]
You need to try 
$ cddexec --redcheck <examples/bug.ine
not 
$ cddexec --rep <examples/bug.ine
which works correctly without the patch.

This exection should return
H-representation
linearity 2  1 2
begin
 2 4 real
  0  0  1  1
 -1  0  1  1
end

which is a correct representation of the empty set.

Komei

**Comment by komeif** [comment]
@dimpase Without the bug fix, 
$ cddexec --redcheck <examples/bug.ine
returns a wrong result (as you pointed out):
H-representation
linearity 1  1
begin
 2 4 real
 -1  0  1  1
 -1  4 -2  0
end

which is not a representation of the empty set.

Komei

**Comment by dimpase** [comment]
OK, this works, thanks.

**Comment by slel** [comment]
Ready to close if this is fixed by #54.

**Comment by kliem** [comment]
It works now. Thank you.
